Mathlib/RingTheory/LinearDisjoint.lean

```diff
@@ -0,0 +1,450 @@
+  Iff.rfl
+
+variable {A B}
+
+@[nontriviality]
+theorem LinearDisjoint.of_subsingleton [Subsingleton R] : A.LinearDisjoint B :=
+  Submodule.LinearDisjoint.of_subsingleton
+
+/-- Linearly disjoint is symmetric if elements in the module commute. -/
+theorem LinearDisjoint.symm_of_commute (H : A.LinearDisjoint B)
+    (hc : ∀ (a : A) (b : B), Commute a.1 b.1) : B.LinearDisjoint A :=
+  Submodule.LinearDisjoint.symm_of_commute H hc
+
+/-- Linearly disjoint is symmetric if elements in the module commute. -/
+theorem linearDisjoint_symm_of_commute
+    (hc : ∀ (a : A) (b : B), Commute a.1 b.1) : A.LinearDisjoint B ↔ B.LinearDisjoint A :=
+  ⟨fun H ↦ H.symm_of_commute hc, fun H ↦ H.symm_of_commute fun _ _ ↦ (hc _ _).symm⟩
+
+namespace LinearDisjoint
+
+variable (A B)
+
+/-- The image of `R` in `S` is linearly disjoint with any other subalgebras. -/
+theorem bot_left : (⊥ : Subalgebra R S).LinearDisjoint B :=
+  Submodule.LinearDisjoint.one_left _
+
+/-- The image of `R` in `S` is linearly disjoint with any other subalgebras. -/
+theorem bot_right : A.LinearDisjoint ⊥ :=
+  Submodule.LinearDisjoint.one_right _
+
+end LinearDisjoint
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R] [CommSemiring S] [Algebra R S]
+
+variable {A B : Subalgebra R S}
+
+/-- Linearly disjoint is symmetric in a commutative ring. -/
+theorem LinearDisjoint.symm (H : A.LinearDisjoint B) : B.LinearDisjoint A :=
+  H.symm_of_commute fun _ _ ↦ mul_comm _ _
+
+/-- Linearly disjoint is symmetric in a commutative ring. -/
+theorem linearDisjoint_symm : A.LinearDisjoint B ↔ B.LinearDisjoint A :=
+  ⟨LinearDisjoint.symm, LinearDisjoint.symm⟩
+
+namespace LinearDisjoint
+
+variable (H : A.LinearDisjoint B)
+include H
+
+/-- If `A` and `B` are subalgebras in a commutative algebra `S` over `R`, and if they are
+linearly disjoint, then there is the natural isomorphism
+`A ⊗[R] B ≃ₐ[R] A ⊔ B` induced by multiplication in `S`. -/
+protected def mulMap :=
+  (AlgEquiv.ofInjective (A.mulMap B) H.injective).trans (equivOfEq _ _ (mulMap_range A B))
+
+@[simp]
+theorem val_mulMap_tmul (a : A) (b : B) : (H.mulMap (a ⊗ₜ[R] b) : S) = a.1 * b.1 := rfl
+
+/-- If `A` and `B` are subalgebras in a commutative algebra `S` over `R`, and if they are
+linearly disjoint, and if they are free `R`-modules, then `A ⊔ B` is also a free `R`-module. -/
+theorem sup_free_of_free [Module.Free R A] [Module.Free R B] : Module.Free R ↥(A ⊔ B) :=
+  Module.Free.of_equiv H.mulMap.toLinearEquiv
+
+end LinearDisjoint
+
+end CommSemiring
+
+section Ring
+
+namespace LinearDisjoint
+
+variable [CommRing R] [Ring S] [Algebra R S]
+
+variable (A B : Subalgebra R S)
+
+lemma mulLeftMap_ker_eq_bot_iff_linearIndependent_op {ι : Type*} (a : ι → A) :
+    LinearMap.ker (Submodule.mulLeftMap (M := toSubmodule A) (toSubmodule B) a) = ⊥ ↔
+    LinearIndependent B.op (MulOpposite.op ∘ A.val ∘ a) := by
+  -- need this instance otherwise `LinearMap.ker_eq_bot` does not work
```

Comment:
The two letI lines can be removed if you start with
```lean
set_option maxSynthPendingDepth 2 in
lemma mulLeftMap_ker_eq_bot_iff_linearIndependent_op ...
```

Reply:
Thank you!